server/src/main/java/org/elasticsearch/health/node/selection/HealthNodeSelectorTaskExecutor.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.health.node.selection;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ResourceAlreadyExistsException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.persistent.AllocatedPersistentTask;
+import org.elasticsearch.persistent.PersistentTaskParams;
+import org.elasticsearch.persistent.PersistentTaskState;
+import org.elasticsearch.persistent.PersistentTasksCustomMetadata;
+import org.elasticsearch.persistent.PersistentTasksExecutor;
+import org.elasticsearch.persistent.PersistentTasksService;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.RemoteTransportException;
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xcontent.ParseField;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.elasticsearch.health.node.selection.HealthNodeSelector.TASK_NAME;
+
+/**
+ * Persistent task executor that is managing the {@link HealthNodeSelector}.
+ */
+public final class HealthNodeSelectorTaskExecutor extends PersistentTasksExecutor<HealthNodeSelectorTaskParams> {
+
+    private static final Logger logger = LogManager.getLogger(HealthNodeSelector.class);
+
+    private final ClusterService clusterService;
+    private final PersistentTasksService persistentTasksService;
+    private final AtomicReference<HealthNodeSelector> currentTask = new AtomicReference<>();
+
+    public HealthNodeSelectorTaskExecutor(Client client, ClusterService clusterService, ThreadPool threadPool) {
+        super(TASK_NAME, ThreadPool.Names.MANAGEMENT);
+        this.clusterService = clusterService;
+        persistentTasksService = new PersistentTasksService(clusterService, threadPool, client);
```

Comment:
We already create a `PersistentTasksService` in `Node` where the `HealthNodeSelectorTaskExecutor` is created. I don't think we should create more than one "service" type thing on a node, so can we pass it in to the constructor instead?

Reply:
No, I think it's an accidental bug and `GeoIpDownloaderTaskExecutor` should have the service passed in instead of creating a new one also.

Reply:
Should I adjust both to avoid future confusions?